### PR TITLE
Various De-Bloating Optimisations

### PR DESCRIPTION
**Pull request by bushrat011899**
# Objective

Reduce the binary size of Bevy applications with negligible negative impact on runtime performance.

## Solution

Using `cargo bloat` I identified methods with disproportionately increased the final binary size of Bevy applications, using the `3d_scene` as my reference for binary sizes. I then refactored these high-value functions to reduce the amount of code present in generic functions, opting to instead move that code into concrete functions. The exact change varies case by case, but the general transformation involves isolating operations requiring the generic parameter and either evaluating them early or providing them as opaque `fn` pointers.

In all cases, public APIs are unaffected.

Currently, this has reduced the final size of the `3d_scene` example by 2.45MiB, 7.3%. I've kept each refactor in its own commit to allow reviewers to see the specific impact each refactor has on the final binary size (please see the commit descriptions). I will also note that some of these may be more controversial than others, and I am open to removing them if an SME has reasonable objections.

On WASM using the `wasm-release` profile, the gains are less extreme, 475KiB or 2.0% saved. However, a size reduction of even that magnitude is still sizable considering the `wasm-release` profile is already optimising for size.

I'm leaving this in a draft state for now while I continue experimenting with refactors. Since these are long-established APIs I don't expect to have much difficulty in keeping this branch updated against `main`.

## Testing

- For each function, `cargo bloat` and my customised analyzer were run to assess the before and after size of the `3d_scene` example to ensure changes positively impacted binary size.
- CI run locally to ensure tests pass.

---

## Showcase

<details>
  <summary>Bloat Analysis Before: 34142KiB</summary>

```
Total Size: 34142KiB
3.26% 1111KiB: core::ops::function::FnOnce::call_once
2.70% 920KiB: hashbrown::raw::inner::RawTable<T,A>::reserve_rehash
1.70% 578KiB: std::sync::once::Once::call_once_force::{{closure}}
1.17% 398KiB: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
1.04% 355KiB: bevy_ecs::query::state::QueryState<D,F>::new_with_access
0.95% 324KiB: bevy_reflect::utility::GenericTypeCell<T>::get_or_insert_by_type_id
0.92% 313KiB: <dyn bevy_reflect::reflect::Reflect>::take
0.89% 303KiB: <bevy_ecs::change_detection::ResMut<T> as bevy_ecs::system::system_param::SystemParam>::init_state
0.88% 299KiB: bevy_ecs::system::system::System::run
0.80% 273KiB: <&mut serde_json::de::Deserializer<R> as serde::de::Deserializer>::deserialize_struct
0.75% 255KiB: <bevy_ecs::change_detection::Res<T> as bevy_ecs::system::system_param::SystemParam>::init_state
0.74% 251KiB: bevy_reflect::struct_trait::StructInfo::new
0.72% 245KiB: bevy_ecs::reflect::from_reflect_with_fallback
0.71% 243KiB: bevy_asset::server::impl$0::load_internal::async_fn$0
0.69% 235KiB: bevy_reflect::type_registry::TypeRegistry::register
0.68% 232KiB: <bevy_ecs::system::function_system::FunctionSystem<Marker,F> as bevy_ecs::system::system::System>::run_unsafe
0.64% 218KiB: <bevy_ecs::system::query::Query<D,F> as bevy_ecs::system::system_param::SystemParam>::init_state
0.60% 203KiB: bevy_reflect::enums::enum_trait::EnumInfo::new
0.56% 190KiB: <bevy_ecs::system::function_system::FunctionSystem<Marker,F> as bevy_ecs::system::system::System>::initialize
0.56% 189KiB: <&T as core::fmt::Debug>::fmt
0.54% 183KiB: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
0.50% 171KiB: bevy_reflect::reflect::PartialReflect::apply
0.49% 167KiB: alloc::sync::Arc<T,A>::drop_slow
0.46% 156KiB: futures_lite::future::block_on
0.46% 156KiB: bevy_ecs::bundle::Bundles::init_info
```

</details>

<details>
  <summary>Bloat Analysis After: 31638KiB</summary>

```
Total Size: 31638KiB
2.75% 870KiB: hashbrown::raw::inner::RawTable<T,A>::reserve_rehash
1.86% 588KiB: std::sync::once::Once::call_once_force::{{closure}}
1.74% 549KiB: core::ops::function::FnOnce::call_once
1.28% 403KiB: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
1.19% 376KiB: bevy_ecs::query::state::QueryState<D,F>::new_with_access
1.06% 336KiB: bevy_reflect::utility::GenericTypeCell<T>::get_or_insert_by_type_id
0.96% 304KiB: bevy_ecs::system::system::System::run
0.86% 273KiB: <&mut serde_json::de::Deserializer<R> as serde::de::Deserializer>::deserialize_struct
0.76% 239KiB: <bevy_ecs::system::query::Query<D,F> as bevy_ecs::system::system_param::SystemParam>::init_state
0.72% 226KiB: <bevy_ecs::system::function_system::FunctionSystem<Marker,F> as bevy_ecs::system::system::System>::run_unsafe
0.67% 212KiB: <dyn bevy_reflect::reflect::Reflect>::take
0.60% 188KiB: <&T as core::fmt::Debug>::fmt
0.58% 184KiB: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
0.58% 182KiB: <bevy_ecs::system::function_system::FunctionSystem<Marker,F> as bevy_ecs::system::system::System>::initialize
0.53% 167KiB: alloc::sync::Arc<T,A>::drop_slow
0.49% 156KiB: <&mut ron::de::Deserializer as serde::de::Deserializer>::deserialize_struct
0.44% 137KiB: bevy_reflect::enums::dynamic_enum::DynamicEnum::from_ref
0.43% 136KiB: bevy_reflect::enums::dynamic_enum::impl$6::try_apply
0.43% 136KiB: alloc::collections::btree::node::Handle<alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Mut,K,V,alloc::collections::btree::node::marker::Leaf>,alloc::collections::btree::node::marker::Edge>::insert_recursing
0.43% 136KiB: bevy_asset::server::impl$0::load_internal_owned::async_fn$0
0.42% 133KiB: hashbrown::raw::RawTable<T,A>::reserve_rehash
0.42% 133KiB: bevy_reflect::reflect::PartialReflect::apply
0.40% 125KiB: futures_lite::future::block_on
0.40% 125KiB: bevy_input::keyboard::_::impl$5::try_apply
0.39% 122KiB: async_task::raw::RawTask<F,T,S,M>::run
```

</details>

### CODE REVIEW

File: crates/bevy_ecs/src/reflect/mod.rs [hunk 1]
```diff
@@ -92,52 +92,61 @@ pub fn from_reflect_with_fallback<T: Reflect + TypePath>(
...
+        TypeId::of::<T>(),
+        std::any::type_name::<T>(),
+    );
+
+    let mut value = value
+        .take::<T>()
+        .unwrap_or_else(|_| different_type_error(reflection_trait, T::type_path()));
 
     value.apply(reflected);
     value
```
Comment (SkiFire13) [inline crates/bevy_ecs/src/reflect/mod.rs:152]: Note that this changes how this function works. Before the result of `ReflectFromReflect` was directly returned, while the others went through the `apply`. Now, all go through the `apply`.

I think you could solve this by moving the `apply` inside the helper method as well, since that can be called on a `dyn Reflect`. This should also further reduce the bloat since you're moving another piece of code from a generic function to a non-generic one.

**Review by SkiFire13 — COMMENTED** [review]
Overall this looks very cool and useful, however I feel like the resulting code patterns might not be familiar to everyone. I would leave some comments quickly explaining that the extra functions are intentional and are supposed to not have generics/have less than the parent function to reduce code bloat.

File: crates/bevy_asset/src/server/mod.rs [hunk 2]
```diff
@@ -494,16 +494,14 @@ impl AssetServer {
     ///
     /// `input_handle` must only be [`Some`] if `should_load` was true when retrieving `input_handle`. This is an optimization to
     /// avoid looking up `should_load` twice, but it means you _must_ be sure a load is necessary when calling this function with [`Some`].
-    async fn load_internal<'a>(
+    async fn load_internal_owned(
         &self,
         mut input_handle: Option<UntypedHandle>,
-        path: AssetPath<'a>,
+        path: AssetPath<'static>,
```
Comment (SludgePhD) [inline crates/bevy_asset/src/server/mod.rs:500]: This change doesn't accomplish anything. Lifetime generics are erased, not monomorphized.

File: crates/bevy_ecs/src/reflect/mod.rs (hunk 1 above)
Comment (bushrat011899) [reply crates/bevy_ecs/src/reflect/mod.rs:152]: Good catch, I missed that!

File: crates/bevy_asset/src/server/mod.rs (hunk 2 above)
Comment (bushrat011899) [reply crates/bevy_asset/src/server/mod.rs:500]: I agree that it _shouldn't_ matter, but it provably reduces the final binary size by 118KiB, more than what I could reasonably chalk up to compiler noise. If you look at [the commit](https://github.com/bevyengine/bevy/pull/14934/commits/b5249249e1597c5d196d6cfffb23aa6f9c4fb32f) in isolation you can see in the description the `cargo bloat` analysis I did before and after just that one change.

I still don't have a thorough explanation as to _why_ this reduces binary size so drastically, since the `'a` lifetime parameter is purely for the borrow checker. My only guess is that because this is an `async fn`, the state machine generation must be dependent on whether this value can be assumed `'static` or not.

Comment (SkiFire13) [reply crates/bevy_asset/src/server/mod.rs:500]: Note that the commit didn't just change a lifetime but also moved a function call outside the function, to another async function, which I think is what actually causes the reduction in size (though I'm not sure **why**, maybe a different allocation pattern for the Future locals?).

After that change the `'static` lifetime is necessary due to some functions being called requiring it, but it shouldn't impact code generation though, since that that stage the compiler already erased all informations about specific lifetimes.

Comment (bushrat011899) [reply crates/bevy_asset/src/server/mod.rs:500]: Yes that is also a valid point, since it shouldn't be possible for the lifetime parameter to contribute to code bloat, it's reasonable to question whether moving that one function call (which also shouldn't contribute to code bloat) is also part of the explanation. I do think this has something to do with how the `async` side of the function works, since I haven't seen any non-async functions with only lifetime parameters contribute to code bloat.

**Comment by bushrat011899** [comment]
After merging in the latest `main`, fixing documentation, and a few other low-hanging fruit, the current bloat analysis is sitting at 31,898KiB, which is higher than a previous commit by just under 200KiB. From experimentation that appears to be the amount of variance observable from a typical set of changes made to Bevy.

Looking at an expanded bloat summary (below), I think we're at a point now where the "easy" wins are already included in this PR. Anything from an external dependency (`std`, `hashbrown`, `core`, etc.) can't really be directly debloated since we'd need to find all the uses of those types, which becomes additionally challenging since they could be indirectly used by Bevy (e.g., `hashbrown::raw::inner::RawTable<T,A>::reserve_rehash` is a private function called as an implementation detail in the `hashbrown` crate). Out of the Bevy controlled functions, the vast majority of the bloat comes from:

- `PartialReflect` and its derived implementation, especially the `try_apply` method.
- `FunctionSystem`
- `Query` and `QueryState`

These items aren't trivial to refactor in a way which removes type parameters, since they're deeply intertwined with their types. I do believe it is possible, especially with the `PartialReflect` trait, but I think it is out of the scope of a general "debloating" PR and should be handled by a more targeted venture.

<details>
  <summary>Current Bloat Analysis: 31898KiB</summary>

```
Total Size: 31898KiB
2.74%    874KiB:         hashbrown::raw::inner::RawTable<T,A>::reserve_rehash
1.85%    591KiB:         std::sync::once::Once::call_once_force::{{closure}}
1.69%    540KiB:         core::ops::function::FnOnce::call_once
1.45%    463KiB:         core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
1.43%    456KiB:         bevy_ecs::query::state::QueryState<D,F>::new_with_access
1.04%    331KiB:         bevy_reflect::utility::GenericTypeCell<T>::get_or_insert_by_type_id
0.94%    300KiB:         bevy_ecs::system::system::System::run
0.86%    273KiB:         <&mut serde_json::de::Deserializer<R> as serde::de::Deserializer>::deserialize_struct
0.67%    212KiB:         <dyn bevy_reflect::reflect::Reflect>::take
0.66%    211KiB:         <bevy_ecs::system::function_system::FunctionSystem<Marker,F> as bevy_ecs::system::system::System>::run_unsafe
0.61%    195KiB:         <bevy_ecs::system::function_system::FunctionSystem<Marker,F> as bevy_ecs::system::system::System>::initialize
0.58%    185KiB:         <&T as core::fmt::Debug>::fmt
0.58%    183KiB:         <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
0.54%    171KiB:         alloc::sync::Arc<T,A>::drop_slow
0.54%    171KiB:         futures_lite::future::block_on
0.51%    163KiB:         <bevy_ecs::system::query::Query<D,F> as bevy_ecs::system::system_param::SystemParam>::init_state
0.44%    138KiB:         bevy_asset::server::impl$0::load_internal_owned::async_fn$0
0.43%    138KiB:         bevy_reflect::enums::dynamic_enum::impl$6::try_apply
0.43%    136KiB:         alloc::collections::btree::node::Handle<alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Mut,K,V,alloc::collections::btree::node::marker::Leaf>,alloc::collections::btree::node::marker::Edge>::insert_recursing
0.42%    133KiB:         hashbrown::raw::RawTable<T,A>::reserve_rehash
0.40%    127KiB:         bevy_ecs::query::state::QueryState<D,F>::new_archetype_internal
0.39%    123KiB:         <&mut ron::de::Deserializer as serde::de::Deserializer>::deserialize_struct
0.37%    119KiB:         async_task::raw::RawTask<F,T,S,M>::run
0.37%    117KiB:         bevy_input::keyboard::_::impl$1::type_info::closure$0
0.36%    116KiB:         bevy_color::color::_::impl$5::try_apply
0.36%    113KiB:         bevy_reflect::reflect::PartialReflect::apply
0.35%    112KiB:         core::slice::sort::merge_sort
0.33%    106KiB:         bevy_ecs::schedule::schedule::ScheduleGraph::process_configs
0.33%    106KiB:         bevy_ecs::intern::Interner<T>::intern
0.31%    99KiB:  hashbrown::raw::inner::RawTableInner::drop_inner_table
0.29%    93KiB:  bevy_reflect::reflect::impl$16::fmt
0.27%    87KiB:  bevy_asset::handle::_::<impl bevy_reflect::reflect::PartialReflect for bevy_asset::handle::Handle<A>>::try_apply
0.27%    85KiB:  core::slice::sort::recurse
0.24%    77KiB:  bevy_asset::id::_::<impl bevy_reflect::reflect::PartialReflect for bevy_asset::id::AssetId<A>>::try_apply
0.24%    77KiB:  <core::panic::unwind_safe::AssertUnwindSafe<F> as core::future::future::Future>::poll
0.24%    76KiB:  wgpu_core::command::bundle::RenderBundleEncoder::finish
0.23%    73KiB:  <bevy_ecs::system::function_system::FunctionSystem<Marker,F> as bevy_ecs::system::system::System>::default_system_sets
0.23%    71KiB:  bevy_ecs::system::function_system::SystemState<Param>::new
0.22%    71KiB:  bevy_gltf::loader::load_gltf::async_fn$0
0.22%    70KiB:  <arrayvec::arrayvec::ArrayVec<T,_> as core::iter::traits::collect::FromIterator<T>>::from_iter
0.21%    68KiB:  <bevy_ecs::system::function_system::FunctionSystem<Marker,F> as bevy_ecs::system::system::System>::update_archetype_component_access
0.21%    66KiB:  hashbrown::map::HashMap<K,V,S,A>::insert
0.21%    65KiB:  taffy::compute::grid::track_sizing::resolve_intrinsic_track_sizes<taffy::tree::taffy_tree::TaffyView<enum2$<bevy_ui::measurement::NodeMeasure>,bevy_ui::layout::ui_surface::impl$2::compute_camera_layout::closure_env$0> >
0.21%    65KiB:  bevy_ecs::world::World::resource_scope
0.20%    65KiB:  bevy_ecs::schedule::config::IntoSystemConfigs::in_set
0.20%    64KiB:  wgpu_core::device::queue::<impl wgpu_core::global::Global>::queue_submit
0.20%    62KiB:  bevy_input::keyboard::_::impl$5::try_apply
0.20%    62KiB:  <&mut ron::de::Deserializer as serde::de::Deserializer>::deserialize_tuple
0.19%    62KiB:  bevy_ecs::schedule::schedule::Schedules::add_systems
0.19%    61KiB:  <&mut serde_json::de::Deserializer<R> as serde::de::Deserializer>::deserialize_map
```
</details>

**Comment by killercup** [comment]
Great effort! Since you only compared the 3d_scene example, I also build all other examples (on macOS with `cargo build --release` with no other config) and here are the sizes in comparison.

**tl;dr** no regression, small improvements between 0-5%

<details>
<summary>Big table</summary>


name | before | after | change
-- | -- | -- | --
change_detection | 1.2M | 1.2M | 100%
events | 1.1M | 1.1M | 100%
resources | 1.1M | 1.1M | 100%
busy_behavior | 620K | 620K | 100%
idle_behavior | 600K | 600K | 100%
generate | 420K | 420K | 100%
2d_gizmos | 64M | 62M | 97%
2d_shapes | 65M | 62M | 95%
2d_top_down_camera | 64M | 62M | 97%
2d_viewport_to_world | 64M | 62M | 97%
3d_gizmos | 64M | 62M | 97%
3d_rotation | 64M | 62M | 97%
3d_scene | 64M | 62M | 97%
3d_shapes | 65M | 62M | 95%
3d_viewport_to_world | 64M | 62M | 97%
alien_cake_addict | 65M | 62M | 95%
align | 64M | 62M | 97%
ambiguity_detection | 64M | 61M | 95%
animate_shader | 65M | 63M | 97%
animated_fox | 64M | 62M | 97%
animated_material | 64M | 62M | 97%
animated_transform | 64M | 62M | 97%
animation_graph | 65M | 62M | 95%
anti_aliasing | 65M | 62M | 95%
array_texture | 66M | 63M | 95%
asset_decompression | 65M | 62M | 95%
asset_loading | 64M | 62M | 97%
asset_settings | 64M | 62M | 97%
async_compute | 64M | 62M | 97%
atmospheric_fog | 64M | 62M | 97%
audio | 64M | 62M | 97%
audio_control | 64M | 62M | 97%
auto_exposure | 65M | 62M | 95%
axes | 64M | 62M | 97%
bevymark | 65M | 62M | 95%
blend_modes | 64M | 62M | 97%
bloom_2d | 64M | 62M | 97%
bloom_3d | 64M | 62M | 97%
borders | 64M | 62M | 97%
bounding_2d | 64M | 62M | 97%
breakout | 64M | 62M | 97%
button | 64M | 62M | 97%
change_detection | 1.2M | 1.2M | 100%
char_input_events | 64M | 62M | 97%
clear_color | 64M | 62M | 97%
color_animation | 64M | 62M | 97%
color_grading | 64M | 62M | 97%
component_hooks | 64M | 62M | 97%
compute_shader_game_of_life | 64M | 62M | 97%
contributors | 64M | 62M | 97%
cubic_curve | 64M | 62M | 97%
cubic_splines | 64M | 62M | 97%
custom_asset | 65M | 62M | 95%
custom_asset_reader | 64M | 62M | 97%
custom_attributes | 608K | 616K | 101%
custom_diagnostic | 64M | 62M | 97%
custom_gltf_vertex_attribute | 65M | 62M | 95%
custom_loop | 1.8M | 1.8M | 100%
custom_phase_item | 65M | 62M | 95%
custom_post_processing | 65M | 62M | 95%
custom_primitives | 64M | 62M | 97%
custom_query_param | 2.0M | 1.9M | 95%
custom_schedule | 1.9M | 1.9M | 100%
custom_skinned_mesh | 64M | 62M | 97%
custom_user_event | 64M | 62M | 97%
custom_vertex_attribute | 66M | 63M | 95%
decodable | 65M | 62M | 95%
deferred_rendering | 65M | 62M | 95%
depth_of_field | 64M | 62M | 97%
desk_toy | 65M | 62M | 95%
display_and_visibility | 64M | 62M | 97%
drag_and_drop | 64M | 61M | 95%
dynamic | 616K | 600K | 97%
dynamic_types | 1.4M | 1.4M | 100%
ecs_guide | 2.0M | 2.0M | 100%
embedded_asset | 64M | 62M | 97%
empty | 1.7M | 1.7M | 100%
empty_defaults | 64M | 61M | 95%
enabling_disabling_diagnostic | 64M | 62M | 97%
event | 64M | 62M | 97%
extended_material | 66M | 63M | 95%
external_source_external_thread | 64M | 62M | 97%
extra_asset_source | 64M | 62M | 97%
fallback_image | 66M | 63M | 95%
first_person_view_model | 64M | 62M | 97%
fixed_timestep | 64M | 62M | 97%
flex_layout | 64M | 62M | 97%
fog | 64M | 62M | 97%
fog_volumes | 64M | 62M | 97%
font_atlas_debug | 64M | 62M | 97%
game_menu | 65M | 62M | 95%
gamepad_input | 64M | 62M | 97%
gamepad_input_events | 64M | 62M | 97%
gamepad_rumble | 64M | 62M | 97%
gamepad_viewer | 64M | 62M | 97%
generate_custom_mesh | 64M | 62M | 97%
generic_reflection | 64M | 62M | 97%
generic_system | 64M | 62M | 97%
gltf_skinned_mesh | 64M | 62M | 97%
gpu_readback | 64M | 62M | 97%
grid | 64M | 62M | 97%
headless | 3.0M | 2.9M | 97%
headless_renderer | 64M | 62M | 97%
hello_world | 1.8M | 1.8M | 100%
hierarchy | 64M | 62M | 97%
irradiance_volumes | 66M | 63M | 95%
iter_combinations | 64M | 62M | 97%
keyboard_input | 64M | 61M | 95%
keyboard_input_events | 64M | 62M | 97%
keyboard_modifiers | 64M | 61M | 95%
light_gizmos | 64M | 62M | 97%
lighting | 64M | 62M | 97%
lightmaps | 64M | 62M | 97%
lines | 65M | 63M | 97%
load_gltf | 64M | 62M | 97%
load_gltf_extras | 64M | 62M | 97%
loading_screen | 65M | 62M | 95%
log_diagnostics | 64M | 62M | 97%
log_layers | 64M | 62M | 97%
log_layers_ecs | 64M | 62M | 97%
logs | 64M | 62M | 97%
low_power | 64M | 62M | 97%
many_animated_sprites | 65M | 62M | 95%
many_buttons | 65M | 62M | 95%
many_cubes | 64M | 62M | 97%
many_foxes | 65M | 62M | 95%
many_gizmos | 64M | 62M | 97%
many_glyphs | 64M | 62M | 97%
many_lights | 64M | 62M | 97%
many_sprites | 64M | 62M | 97%
mesh2d | 64M | 62M | 97%
mesh2d_alpha_mode | 64M | 62M | 97%
mesh2d_arcs | 64M | 62M | 97%
mesh2d_manual | 64M | 62M | 97%
mesh2d_vertex_color_texture | 64M | 62M | 97%
minimising | 64M | 62M | 97%
monitor_info | 64M | 62M | 97%
morph_targets | 64M | 62M | 97%
motion_blur | 65M | 62M | 95%
mouse_grab | 64M | 62M | 97%
mouse_input | 64M | 62M | 97%
mouse_input_events | 64M | 62M | 97%
move_sprite | 64M | 62M | 97%
multi_asset_sync | 65M | 62M | 95%
multiple_windows | 64M | 62M | 97%
no_prepass | 64M | 61M | 95%
no_renderer | 64M | 61M | 95%
nondeterministic_system_order | 64M | 62M | 97%
observe | 64M | 62M | 97%
observer_propagation | 6.1M | 6.0M | 98%
observers | 64M | 62M | 97%
one_shot_systems | 64M | 62M | 97%
orthographic | 64M | 62M | 97%
overflow | 64M | 62M | 97%
overflow_debug | 65M | 62M | 95%
parallax_mapping | 64M | 62M | 97%
parallel_query | 64M | 62M | 97%
parenting | 64M | 62M | 97%
pbr | 64M | 62M | 97%
physics_in_fixed_timestep | 64M | 62M | 97%
pitch | 64M | 62M | 97%
pixel_grid_snap | 64M | 62M | 97%
plugin | 64M | 62M | 97%
plugin_group | 64M | 62M | 97%
post_processing | 64M | 62M | 97%
random_sampling | 64M | 62M | 97%
reflection | 64M | 62M | 97%
reflection_probes | 65M | 62M | 95%
reflection_types | 64M | 62M | 97%
relative_cursor_position | 64M | 62M | 97%
removal_detection | 64M | 62M | 97%
render_primitives | 65M | 62M | 95%
render_to_texture | 64M | 62M | 97%
render_ui_to_texture | 64M | 62M | 97%
repeated_texture | 64M | 62M | 97%
resizing | 64M | 62M | 97%
return_after_run | 64M | 62M | 97%
rotation | 64M | 62M | 97%
run_conditions | 64M | 62M | 97%
sampling_primitives | 64M | 62M | 97%
scale | 64M | 62M | 97%
scale_factor_override | 64M | 62M | 97%
scene | 65M | 62M | 95%
scene_viewer | 65M | 62M | 95%
screenshot | 64M | 62M | 97%
scrolling_fog | 65M | 62M | 95%
send_and_receive_events | 3.1M | 3.0M | 97%
shader_defs | 65M | 63M | 97%
shader_instancing | 65M | 62M | 95%
shader_material | 66M | 63M | 95%
shader_material_2d | 65M | 62M | 95%
shader_material_screenspace_texture | 66M | 63M | 95%
shader_prepass | 66M | 63M | 95%
shadow_biases | 64M | 62M | 97%
shadow_caster_receiver | 64M | 62M | 97%
simple_picking | 64M | 62M | 97%
size_constraints | 64M | 62M | 97%
skybox | 64M | 62M | 97%
smooth_follow | 64M | 62M | 97%
soundtrack | 64M | 62M | 97%
spatial_audio_2d | 64M | 62M | 97%
spatial_audio_3d | 64M | 62M | 97%
specialized_mesh_pipeline | 65M | 62M | 95%
spherical_area_lights | 64M | 62M | 97%
split_screen | 64M | 62M | 97%
spotlight | 64M | 62M | 97%
sprite | 64M | 62M | 97%
sprite_animation | 64M | 62M | 97%
sprite_flipping | 64M | 62M | 97%
sprite_picking | 64M | 62M | 97%
sprite_sheet | 64M | 62M | 97%
sprite_slice | 64M | 62M | 97%
sprite_tile | 64M | 62M | 97%
ssao | 64M | 62M | 97%
ssr | 66M | 63M | 95%
startup_system | 1.8M | 1.8M | 100%
system_closure | 4.7M | 4.7M | 100%
system_param | 1.9M | 1.9M | 100%
system_piping | 4.7M | 4.7M | 100%
text | 64M | 62M | 97%
text2d | 64M | 62M | 97%
text_debug | 64M | 62M | 97%
text_input | 64M | 62M | 97%
text_pipeline | 64M | 62M | 97%
text_wrap_debug | 64M | 62M | 97%
texture | 64M | 62M | 97%
texture_atlas | 65M | 62M | 95%
texture_binding_array | 66M | 63M | 95%
thread_pool_resources | 64M | 61M | 95%
time | 3.0M | 3.0M | 100%
timers | 64M | 62M | 97%
tonemapping | 66M | 63M | 95%
touch_input | 64M | 62M | 97%
touch_input_events | 64M | 61M | 95%
transform | 64M | 62M | 97%
transform_hierarchy | 64M | 62M | 97%
translation | 64M | 62M | 97%
transmission | 65M | 62M | 95%
transparency_2d | 64M | 62M | 97%
transparency_3d | 64M | 62M | 97%
transparency_ui | 64M | 62M | 97%
transparent_window | 64M | 62M | 97%
two_passes | 64M | 62M | 97%
type_data | 968K | 968K | 100%
ui | 64M | 62M | 97%
ui_material | 65M | 62M | 95%
ui_scaling | 64M | 62M | 97%
ui_texture_atlas | 64M | 62M | 97%
ui_texture_atlas_slice | 64M | 62M | 97%
ui_texture_slice | 64M | 62M | 97%
update_gltf_scene | 64M | 62M | 97%
vertex_colors | 64M | 62M | 97%
viewport_debug | 64M | 62M | 97%
virtual_time | 64M | 62M | 97%
visibility_range | 64M | 62M | 97%
volumetric_fog | 64M | 62M | 97%
window_fallthrough | 64M | 62M | 97%
window_resizing | 64M | 62M | 97%
window_settings | 64M | 62M | 97%
wireframe | 65M | 62M | 95%
wireframe_2d | 65M | 62M | 95%
without_winit | 64M | 62M | 97%
z_index | 64M | 62M | 97%

</details>

(I ran `cargo metadata -q --no-deps | jq -r '.packages[].targets[] | select(.kind[] == "example") | .name' | xargs -I {} sh -c '[ -e "$0/$1" ] && echo "$1 | $(du -sh "$0/$1" | cut -f1)"' "$CARGO_TARGET_DIR/release/examples/" "{}"` twice and manually merged the output and processed it with a bit of python)

**Comment by bushrat011899** [comment]
Wow that's brilliant, thanks! On my laptop it took about 10 minutes per `cargo bloat` invocation so I hadn't considered a wider investigation as feasible. I was personally hoping to get a larger improvement, but even an (up to) 5% binary reduction seems worth this one-time effort.

**Comment by bushrat011899** [comment]
Closing this as _having missed the train_. Not a particularly high priority change, but the techniques are documented here if anyone wishes to try again!
